File: frontend/src/lib/ntdll.js
/**
 * Cliente da ntdll.dll importada.
 *
 * A tabela e VOLATIL no servidor (vive em memoria, some quando o container
 * reinicia), e aqui ela e mantida num registro de modulo — carregada uma vez e
 * consultada a cada passo da execucao, sem ida e volta ao backend.
 */

import api from "lib/api";

/** SSN -> nome, por arquitetura. Vazio ate alguem importar. */
const TABLES = { x86: null, x86_64: null };

/** Status por arquitetura: `{ count, origin }`. */
const SUMMARIES = { x86: null, x86_64: null };

export function ntdllTable(arch) {
  return TABLES[arch] || null;
}

export function ntdllSummary(arch) {
  return SUMMARIES[arch] || null;
}

/** Nome da funcao para um SSN, ou null se nao houver tabela. */
export function resolveSyscall(arch, ssn) {
  const table = TABLES[arch];
  if (!table) return null;
  return table[String(ssn)] || null;
}

function adopt(arch, summary, syscalls) {
  TABLES[arch] = syscalls || null;
  SUMMARIES[arch] = summary || null;
}

/** Consulta o servidor e adota o que ja estiver carregado. */
export async function refreshNtdll(arch) {
  const { data } = await api.get("/api/ntdll/", { params: { arch } });
  const summary = data.loaded?.[arch] || null;
  adopt(arch, summary, summary ? data.syscalls : null);
  return summary;
}

/** Envia a DLL; o servidor extrai os SSN e devolve a tabela. */
export async function importNtdll(file) {
  const form = new FormData();
  form.append("file", file);
  const { data } = await api.post("/api/ntdll/", form);
  adopt(data.arch, { count: data.count, origin: data.origin, exports: data.exports },
        data.syscalls);
  return data;
}

/** Descarta a tabela, no servidor e aqui. */
export async function clearNtdll(arch) {
  await api.delete("/api/ntdll/", { params: arch ? { arch } : undefined });
  if (arch) adopt(arch, null, null);
  else {
    adopt("x86", null, null);
    adopt("x86_64", null, null);
  }
}

File: backend/asm_simulator/views/ntdll.py
"""Import da ``ntdll.dll`` para resolver numeros de syscall do Windows.

A tabela e VOLATIL: vive num diretorio em memoria e some quando o container
reinicia. Ver ``services/ntdll.py`` para o porque.
"""

import logging

from rest_framework import status
from rest_framework.response import Response
from rest_framework.views import APIView

from asm_simulator.i18n import tr
from asm_simulator.services import ntdll

log = logging.getLogger(__name__)

SUPPORTED_ARCHS = ('x86', 'x86_64')

ERROR_TEXTS = {
    'ntdll.notPe': 'This file is not a Windows DLL.',
    'ntdll.unsupportedMachine': 'This DLL is not x86 or x86-64.',
    'ntdll.noExports': 'This DLL has no export table.',
    'ntdll.noSyscalls': 'No syscall stubs found — is this really ntdll.dll?',
    'ntdll.noStorage': 'No writable memory storage available.',
    'ntdll.missing': 'Choose an ntdll.dll to import.',
    'ntdll.tooLarge': 'This file is too large.',
}


def _summary(table):
    """O que a interface precisa saber. A tabela inteira so na leitura."""
    if not table:
        return None
    return {
        'arch': table['arch'],
        'count': table['count'],
        'exports': table.get('exports'),
        'origin': table.get('origin'),
    }


class NtdllView(APIView):
    """GET status; POST importa; DELETE descarta."""

    def get(self, request):
        arch = request.query_params.get('arch')
        archs = [arch] if arch in SUPPORTED_ARCHS else SUPPORTED_ARCHS

        loaded = {}
        for item in archs:
            table = ntdll.load(item)
            if table:
                loaded[item] = _summary(table)

        # `syscalls` so vai quando se pede UMA arquitetura: e a tabela inteira,
        # e a interface a carrega uma vez para resolver sem ida e volta a cada
        # passo da execucao.
        body = {'loaded': loaded}
        if arch in SUPPORTED_ARCHS:
            table = ntdll.load(arch)
            body['syscalls'] = table['syscalls'] if table else {}
        return Response(body)

    def post(self, request):
        upload = request.FILES.get('file')
        if upload is None:
            return self._error(request, 'ntdll.missing')
        if upload.size > ntdll.MAX_DLL_BYTES:
            return self._error(request, 'ntdll.tooLarge')

        try:
            table = ntdll.parse(upload.read())
        except ntdll.NtdllError as exc:
            return self._error(request, str(exc))

        stored = ntdll.store(table, origin=upload.name)
        return Response(
            {**_summary(stored), 'syscalls': stored['syscalls']},
            status=status.HTTP_201_CREATED,
        )

    def delete(self, request):
        arch = request.query_params.get('arch')
        ntdll.clear(arch if arch in SUPPORTED_ARCHS else None)
        return Response(status=status.HTTP_204_NO_CONTENT)

    def _error(self, request, key):
        return Response(
            {'detail': tr(request, key, ERROR_TEXTS.get(key, 'Invalid file.'))},
            status=status.HTTP_400_BAD_REQUEST,
        )
